tests/optional_kernel_features/kernel_features_common.h

```diff
@@ -0,0 +1,519 @@
+              cgh.parallel_for(sycl::range{1},                                 \
+                               [=](sycl::item<1>) ATTRIBUTE { __VA_ARGS__; }); \
+            })                                                                 \
+            .wait();                                                           \
+        is_exception_thrown = false;                                           \
+      } catch (const sycl::exception &e) {                                     \
+        is_exception_thrown = true;                                            \
+        if (IS_EXCEPTION_EXPECTED) {                                           \
+          INFO("Error code check");                                            \
+          CHECK(e.code() == ERRC);                                             \
+        }                                                                      \
+      }                                                                        \
+      CHECK(is_exception_thrown == IS_EXCEPTION_EXPECTED);                     \
+      check_async_exception(QUEUE, false);                                     \
+    }                                                                          \
+    {                                                                          \
+      INFO("Execution of submission call in parallel_for_work_group");         \
+      try {                                                                    \
+        QUEUE                                                                  \
+            .submit([&](sycl::handler &cgh) {                                  \
+              cgh.parallel_for_work_group(sycl::range{1}, sycl::range{1},      \
+                                          [=](sycl::group<1>)                  \
+                                              ATTRIBUTE { __VA_ARGS__; });     \
+            })                                                                 \
+            .wait();                                                           \
+        is_exception_thrown = false;                                           \
+      } catch (const sycl::exception &e) {                                     \
+        is_exception_thrown = true;                                            \
+        if (IS_EXCEPTION_EXPECTED) {                                           \
+          INFO("Error code check");                                            \
+          CHECK(e.code() == ERRC);                                             \
+        }                                                                      \
+      }                                                                        \
+      CHECK(is_exception_thrown == IS_EXCEPTION_EXPECTED);                     \
+      check_async_exception(QUEUE, false);                                     \
+    }                                                                          \
```

Comment:
Why such a huge macro instead of a function/class/lambda? Is this to pass the attribute with the preprocessor?
Also a lot of copy-paste inside the macro itself.

Reply:
I used macro because in some cases kernel should be sent as a submission call (without storing it in separate lambda or class functor for example). To achieve that I wrote this huge macro, that should paste code with submission call.

This macro also refactored in this commit: 2fb2d5e4004329d31498a25c27355e0fc3c59dd7